gbpservice/neutron/services/servicechain/plugins/ncp/node_drivers/nfp_node_driver.py

```diff
@@ -0,0 +1,699 @@
+            plumber_base.SERVICE_TARGET_NAME_PREFIX) or
+            policy_target['name'].startswith('tscp_endpoint_service') or
+            policy_target['name'].startswith('vip_pt')):
+            return True
+        else:
+            return False
+
+    def _resource_owner_tenant_id(self):
+        user, pwd, tenant, auth_url = utils.get_keystone_creds()
+        keystoneclient = keyclient.Client(username=user, password=pwd,
+                                          auth_url=auth_url)
+        try:
+            tenant = keystoneclient.tenants.find(name=tenant)
+            return tenant.id
+        except k_exceptions.NotFound:
+            with excutils.save_and_reraise_exception(reraise=True):
+                LOG.error(_('No tenant with name %s exists.'), tenant)
+        except k_exceptions.NoUniqueMatch:
+            with excutils.save_and_reraise_exception(reraise=True):
+                LOG.error(_('Multiple tenants matches found for %s'), tenant)
+
+    def _get_resource_owner_context(self, plugin_context):
+        if cfg.CONF.nfp_node_driver.is_service_admin_owned:
+            resource_owner_context = plugin_context.elevated()
+            resource_owner_context.tenant_id = self.resource_owner_tenant_id
+            user, pwd, ignore_tenant, auth_url = utils.get_keystone_creds()
+            keystoneclient = keyclient.Client(username=user, password=pwd,
+                                              auth_url=auth_url)
+            resource_owner_context.auth_token = keystoneclient.get_token(
+                self.resource_owner_tenant_id)
+            return resource_owner_context
+        else:
+            return plugin_context
+
+    def _update(self, context, pt_added_or_removed=False):
+        if context.current_profile['service_type'] == pconst.LOADBALANCER:
+            if (not context.original_node or
+                context.original_node == context.current_node):
+                LOG.info(_("No action to take on update"))
+                return
+        self.nfp_notifier.update_service_config()
+
+    def _get_management_ptg_id(self, context):
+        # REVISIT(Magesh): Retrieving management PTG by name will not be
+        # required when the service_ptg patch is merged
+        filters = {'name': [SVC_MGMT_PTG_NAME]}
+        svc_mgmt_ptgs = context.gbp_plugin.get_policy_target_groups(
+            context.plugin_context, filters)
+        if not svc_mgmt_ptgs:
+            LOG.error(_("Service Management Group is not created by Admin"))
+            raise Exception()
+        else:
+            return svc_mgmt_ptgs[0]['id']
+
+    def _get_service_target_from_relations(self, context, relationship):
+        LOG.debug("Relation: %s instance_id: %s node_id: %s tenant_id: %s"
+                  % (relationship, context.instance['id'],
+                     context.current_node['id'],
+                     context._plugin_context.tenant_id))
+        service_targets = model.get_service_targets(
```

Comment:
Is this for retrieving the ports created by the plumber? Does the plumber associated these with the sc_inst_id:sc_node_id?